File: BinarySearchTree.java

public class BinarySearchTree {

    // Node class
    static class Node {

        int data;
        Node left;
        Node right;

        // Constructor
        Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }

    }

    // Root node
    Node root;

    // Constructor
    public BinarySearchTree() {
        root = null;
    }

    // Insert a new node
    public Node insert(Node node, int data) {

        // If the tree is empty, create a new node
        if (node == null) {
            return new Node(data);
        }

        // Insert into the left subtree
        if (data < node.data) {
            node.left = insert(node.left, data);
        }

        // Insert into the right subtree
        else if (data > node.data) {
            node.right = insert(node.right, data);
        }

        return node;
    }

    // Search for a value
    public boolean search(Node node, int key) {

        // Value not found
        if (node == null) {
            return false;
        }

        // Value found
        if (node.data == key) {
            return true;
        }

        // Search left subtree
        if (key < node.data) {
            return search(node.left, key);
        }

        // Search right subtree
        return search(node.right, key);

    }

    // Inorder Traversal (Left -> Root -> Right)
    public void inorder(Node node) {

        if (node == null) {
            return;
        }

        inorder(node.left);

        System.out.print(node.data + " ");

        inorder(node.right);

    }

    public static void main(String[] args) {

        // Create Binary Search Tree
        BinarySearchTree bst = new BinarySearchTree();

        // Insert nodes
        bst.root = bst.insert(bst.root, 50);
        bst.root = bst.insert(bst.root, 30);
        bst.root = bst.insert(bst.root, 70);
        bst.root = bst.insert(bst.root, 20);
        bst.root = bst.insert(bst.root, 40);
        bst.root = bst.insert(bst.root, 60);
        bst.root = bst.insert(bst.root, 80);

        // Print tree in sorted order
        System.out.print("Inorder Traversal: ");
        bst.inorder(bst.root);

        System.out.println();

        // Search for elements
        System.out.println("Search 40: " + bst.search(bst.root, 40));
        System.out.println("Search 90: " + bst.search(bst.root, 90));

    }

}

File: BinarySearchTree.c
#include <stdio.h>
#include <stdlib.h>

// Node Structure
struct Node {
    int data;
    struct Node *left;
    struct Node *right;
};

// Create a New Node
struct Node* createNode(int value) {

    struct Node *newNode = (struct Node*)malloc(sizeof(struct Node));

    newNode->data = value;
    newNode->left = NULL;
    newNode->right = NULL;

    return newNode;
}

// Insert Node
struct Node* insert(struct Node *root, int value) {

    if(root == NULL) {
        return createNode(value);
    }

    if(value < root->data) {
        root->left = insert(root->left, value);
    }
    else if(value > root->data) {
        root->right = insert(root->right, value);
    }

    return root;
}

// Search Node
struct Node* search(struct Node *root, int key) {

    if(root == NULL || root->data == key) {
        return root;
    }

    if(key < root->data) {
        return search(root->left, key);
    }

    return search(root->right, key);
}

// Inorder Traversal
void inorder(struct Node *root) {

    if(root == NULL)
        return;

    inorder(root->left);
    printf("%d ", root->data);
    inorder(root->right);
}

// Preorder Traversal
void preorder(struct Node *root) {

    if(root == NULL)
        return;

    printf("%d ", root->data);
    preorder(root->left);
    preorder(root->right);
}

// Postorder Traversal
void postorder(struct Node *root) {

    if(root == NULL)
        return;

    postorder(root->left);
    postorder(root->right);
    printf("%d ", root->data);
}

int main() {

    struct Node *root = NULL;

    root = insert(root, 50);
    root = insert(root, 30);
    root = insert(root, 70);
    root = insert(root, 20);
    root = insert(root, 40);
    root = insert(root, 60);
    root = insert(root, 80);

    printf("Inorder Traversal:\n");
    inorder(root);

    printf("\n\n");

    printf("Preorder Traversal:\n");
    preorder(root);

    printf("\n\n");

    printf("Postorder Traversal:\n");
    postorder(root);

    printf("\n\n");

    int key = 60;

    if(search(root, key) != NULL)
        printf("%d found in the BST.\n", key);
    else
        printf("%d not found in the BST.\n", key);

    return 0;
}
